[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github.com/505539367/AI-class/blob/main/Assignment%204%20-%20Ying%20Yang.ipynb)

In [1]:
# Imports and pip installations (if needed)
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
import warnings
warnings.filterwarnings("ignore")

## Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
dataset = load_iris()
dataframe = pd.DataFrame(data = dataset['data'],
                     columns= dataset['feature_names'])
dataframe["target"] = dataset.target
# Output the first 15 rows of the data
dataframe.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  
10       0  
11       0  
12       0  
13       0  
14       0

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
print(dataset.target_names)

['setosa' 'versicolor' 'virginica']


In [4]:
print(dataset.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [5]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


In [6]:
dataframe.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

### About the dataset
There are 150 samples from the iris dataset.
Included three species of Iris:

    0 represent Iris-Setosa
    1 represent Iris-Versicolour
    2 represent Iris-Virginica

The features should be dataset['data'], which contains sepal length, sepal width, petal length, petal width.

The label will be target, which has value of 0, 1, and 2. The value is representing a species of Iris.




## Part 2: Split the dataset into train and test

In [7]:
# Take the dataset and split it into our features (X) and label (y)
X = dataframe[dataset.feature_names]
X #Features

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [8]:
y = dataframe["target"]
y #labels

0      0
1      0
2      0
3      0
4      0
      ..
145    2
146    2
147    2
148    2
149    2
Name: target, Length: 150, dtype: int32

In [9]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1)
X_train

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
60                 5.0               2.0                3.5               1.0
127                6.1               3.0                4.9               1.8
49                 5.0               3.3                1.4               0.2
101                5.8               2.7                5.1               1.9
90                 5.5               2.6                4.4               1.2
..                 ...               ...                ...               ...
117                7.7               3.8                6.7               2.2
68                 6.2               2.2                4.5               1.5
18                 5.7               3.8                1.7               0.3
97                 6.2               2.9                4.3               1.3
44                 5.1               3.8                1.9               0.4

[135 rows x 4 columns]

In [10]:
X_test

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
98                 5.1               2.5                3.0               1.1
54                 6.5               2.8                4.6               1.5
6                  4.6               3.4                1.4               0.3
80                 5.5               2.4                3.8               1.1
107                7.3               2.9                6.3               1.8
74                 6.4               2.9                4.3               1.3
39                 5.1               3.4                1.5               0.2
34                 4.9               3.1                1.5               0.2
79                 5.7               2.6                3.5               1.0
125                7.2               3.2                6.0               1.8
73                 6.1               2.8                4.7               1.2
132                6.4               2.8                5.6               2.2
137                6.4               3.1                5.5               1.8
62                 6.0               2.2                4.0               1.0
105                7.6               3.0                6.6               2.1

## Part 3: Logistic Regression

In [11]:
# i. Use sklearn to train a LogisticRegression model on the training set
LR_model = LogisticRegression()
LR_model.fit(X_train, y_train)

LogisticRegression()

In [12]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X[:1]
sample_pred = LR_model.predict_proba(sample_point)[0]

print("Class\t\tSpecies\t\t\tProbability")
print("0\t\tSetosa\t\t\t{:0.5f}".format(sample_pred[0]))
print("1\t\tVersicolor\t\t{:0.5f}".format(sample_pred[1]))
print("2\t\tVirginica\t\t{:0.5f}".format(sample_pred[2]))

Class		Species			Probability
0		Setosa			0.98187
1		Versicolor		0.01813
2		Virginica		0.00000


In [13]:
#Predict the x_test data set, and get the probabilities for each possible class
print("y_test\t\tProbability of Setosa\tProbability of Versicolor\tProbability of Virginica",end='')
index = 0;
for i in LR_model.predict(X_test):
    print("\n",i,end ='')
    for j in LR_model.predict_proba(X_test)[index]:
        print("\t\t\t{:0.5f}".format(j),end ='')
    index+= 1;

y_test		Probability of Setosa	Probability of Versicolor	Probability of Virginica
 1			0.29656			0.70179			0.00164
 1			0.00324			0.80018			0.19658
 0			0.98698			0.01302			0.00000
 1			0.03913			0.94417			0.01671
 2			0.00000			0.02476			0.97524
 1			0.00934			0.93418			0.05648
 0			0.97465			0.02535			0.00000
 0			0.97057			0.02943			0.00000
 1			0.07955			0.91615			0.00430
 2			0.00001			0.05876			0.94124
 1			0.00398			0.85115			0.14487
 2			0.00001			0.02983			0.97016
 2			0.00010			0.12972			0.87018
 1			0.01293			0.96904			0.01803
 2			0.00000			0.00568			0.99432

All the prediction are correct in the result above.

In [14]:
# iii. Report on the score for Logistic regression model, what does the score measure?
train_score = LR_model.score(X_train, y_train)
test_score  = LR_model.score(X_test, y_test)
print("Training score for model :", train_score)
print("Test score for model :", test_score)

Training score for model : 0.9703703703703703
Test score for model : 1.0


### Score:
The score of the model is between 0 to 1 (0% to 100%) and it represent the accuracy on the given data. 
the score of training data is 0.97, means the accuracy on the model is about 97%.
The score of test data is 1.0, means model is perfectly accurate.

In [15]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
print("Coeffients :", LR_model.coef_) 
print("Intercept :", LR_model.intercept_) 

Coeffients : [[-0.40638851  0.88453867 -2.4644257  -1.00915946]
 [ 0.49881654 -0.23003799 -0.15371162 -0.96316236]
 [-0.09242803 -0.65450068  2.61813732  1.97232182]]
Intercept : [  9.80962203   1.85815545 -11.66777747]


## Part 4: Support Vector Machine

In [16]:
# i. Use sklearn to train a Support Vector Classifier on the training set
SVM_model = SVC(probability = True)
SVM_model.fit(X_train, y_train)

SVC(probability=True)

In [17]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X[0:1]
print(sample_point)
sample_pred = SVM_model.predict_proba(sample_point)[0]

print("\nClass\t\tSpecies\t\t\tProbability")
print("0\t\tSetosa\t\t\t{:0.5f}".format(sample_pred[0]))
print("1\t\tVersicolor\t\t{:0.5f}".format(sample_pred[1]))
print("2\t\tVirginica\t\t{:0.5f}".format(sample_pred[2]))

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2

Class		Species			Probability
0		Setosa			0.97795
1		Versicolor		0.01363
2		Virginica		0.00842


In [18]:
#Predict the x_test data set, and get the probabilities for each possible class
print("y_test\t\tProbability of Setosa\tProbability of Versicolor\tProbability of Virginica",end='')
index = 0;
for i in SVM_model.predict(X_test):
    print("\n",i,end ='')
    for j in SVM_model.predict_proba(X_test)[index]:
        print("\t\t\t{:0.5f}".format(j),end ='')
    index+= 1;

y_test		Probability of Setosa	Probability of Versicolor	Probability of Virginica
 1			0.11994			0.85765			0.02241
 1			0.00786			0.87947			0.11267
 0			0.97716			0.01409			0.00875
 1			0.01359			0.98164			0.00477
 2			0.00947			0.00226			0.98827
 1			0.01040			0.97634			0.01326
 0			0.97277			0.01794			0.00929
 0			0.96771			0.02243			0.00986
 1			0.03559			0.95617			0.00823
 2			0.00924			0.00671			0.98405
 1			0.00671			0.92662			0.06667
 2			0.00682			0.00359			0.98959
 2			0.00834			0.03333			0.95834
 1			0.00899			0.98640			0.00461
 2			0.01202			0.00150			0.98647

All the predictions are correct in the result above

In [19]:
# iii. Report on the score for the SVM, what does the score measure?
train_score = SVM_model.score(X_train, y_train)
test_score  = SVM_model.score(X_test, y_test)
print("Training score for model :", train_score)
print("Test score for model :", test_score)

Training score for model : 0.9703703703703703
Test score for model : 1.0


### Score:
The score of the model is between 0 to 1 (0% to 100%) and it represent the accuracy on the given data. The score of training data is 0.97, means the accuracy on the model is about 97%. The score of test data is 1.0, means model is perfectly accurate.

## Part 5: Neural Network

In [37]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
NN_model = MLPClassifier(max_iter = 100)
NN_model.fit(X_train, y_train)

MLPClassifier(max_iter=100)

In [38]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X[0:1]
print(sample_point)
sample_pred = NN_model.predict_proba(sample_point)[0]

print("\nClass\t\tSpecies\t\t\tProbability")
print("0\t\tSetosa\t\t\t{:0.5f}".format(sample_pred[0]))
print("1\t\tVersicolor\t\t{:0.5f}".format(sample_pred[1]))
print("2\t\tVirginica\t\t{:0.5f}".format(sample_pred[2]))


   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2

Class		Species			Probability
0		Setosa			0.84199
1		Versicolor		0.12079
2		Virginica		0.03721


In [39]:
#Predict the x_test data set, and get the probabilities for each possible class
print("y_test\t\tProbability of Setosa\tProbability of Versicolor\tProbability of Virginica",end='')
index = 0;
for i in NN_model.predict(X_test):
    print("\n",i,end ='')
    for j in NN_model.predict_proba(X_test)[index]:
        print("\t\t\t{:0.5f}".format(j),end ='')
    index+= 1;

y_test		Probability of Setosa	Probability of Versicolor	Probability of Virginica
 1			0.23384			0.44782			0.31835
 2			0.08052			0.45777			0.46171
 0			0.82237			0.12977			0.04785
 1			0.13286			0.46938			0.39776
 2			0.02131			0.35473			0.62396
 1			0.12207			0.47374			0.40418
 0			0.81888			0.13718			0.04393
 0			0.77365			0.16923			0.05712
 1			0.19828			0.47137			0.33035
 2			0.03386			0.37850			0.58763
 2			0.09674			0.44371			0.45955
 2			0.02315			0.33729			0.63956
 2			0.04368			0.37565			0.58067
 1			0.10871			0.49272			0.39857
 2			0.01459			0.32749			0.65792

All the predictions are correct in the result above

In [40]:
# iii. Report on the score for the Neural Network, what does the score measure?
train_score = NN_model.score(X_train, y_train)
test_score  = NN_model.score(X_test, y_test)
print("Training score for model :", train_score)
print("Test score for model :", test_score)

Training score for model : 0.8740740740740741
Test score for model : 0.8666666666666667


### Score:
The score of the model is between 0 to 1 (0% to 100%) and it represent the accuracy on the given data. The score of training data is 0.87, means the accuracy on the model is about 87%. The score of test data is 0.866, means model is 87% accurate.

In [53]:
# iv: Experiment with different options for the neural network, report on your best configuration
#(the highest score I was able to achieve was 0.8666)
print("When max_iter = 300:")
NN_model = MLPClassifier(max_iter = 200)
NN_model.fit(X_train, y_train)
train_score = NN_model.score(X_train, y_train)
test_score  = NN_model.score(X_test, y_test)
print("\tTraining score for model :", train_score)
print("\tTest score for model :", test_score)

When max_iter = 300:
	Training score for model : 0.9629629629629629
	Test score for model : 1.0


In [59]:
print("When max_iter = 300:")
NN_model = MLPClassifier(max_iter = 300)
NN_model.fit(X_train, y_train)
train_score = NN_model.score(X_train, y_train)
test_score  = NN_model.score(X_test, y_test)
print("\tTraining score for model :", train_score)
print("\tTest score for model :", test_score)

When max_iter = 300:
	Training score for model : 0.9777777777777777
	Test score for model : 1.0


In [76]:
print("When default configuration")
NN_model = MLPClassifier()
NN_model.fit(X_train, y_train)
train_score = NN_model.score(X_train, y_train)
test_score  = NN_model.score(X_test, y_test)
print("\tTraining score for model :", train_score)
print("\tTest score for model :", test_score)

When default configuration
	Training score for model : 0.9703703703703703
	Test score for model : 1.0


When max_iter change from 100 to 200:

    The training score increasing from 0.87 to 0.96.
    The test score increasing from 0.87 to 1.0

When max_iter change from 100 to 200:

    The training score increasing from 0.87 to 0.97.
    The test score increasing from 0.87 to 1.0
    

## Part 6: K-Nearest Neighbors

In [67]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
KNN_model = KNeighborsClassifier()
KNN_model.fit(X_train, y_train)

KNeighborsClassifier()

In [70]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_point = X[0:1]
print(sample_point)
sample_pred = KNN_model.predict_proba(sample_point)[0]

print("\nClass\t\tSpecies\t\t\tProbability")
print("0\t\tSetosa\t\t\t{:0.5f}".format(sample_pred[0]))
print("1\t\tVersicolor\t\t{:0.5f}".format(sample_pred[1]))
print("2\t\tVirginica\t\t{:0.5f}".format(sample_pred[2]))

   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2

Class		Species			Probability
0		Setosa			1.00000
1		Versicolor		0.00000
2		Virginica		0.00000


In [71]:
#Predict the x_test data set, and get the probabilities for each possible class
print("y_test\t\tProbability of Setosa\tProbability of Versicolor\tProbability of Virginica",end='')
index = 0;
for i in KNN_model.predict(X_test):
    print("\n",i,end ='')
    for j in KNN_model.predict_proba(X_test)[index]:
        print("\t\t\t{:0.5f}".format(j),end ='')
    index+= 1;

y_test		Probability of Setosa	Probability of Versicolor	Probability of Virginica
 1			0.00000			1.00000			0.00000
 1			0.00000			1.00000			0.00000
 0			1.00000			0.00000			0.00000
 1			0.00000			1.00000			0.00000
 2			0.00000			0.00000			1.00000
 1			0.00000			1.00000			0.00000
 0			1.00000			0.00000			0.00000
 0			1.00000			0.00000			0.00000
 1			0.00000			1.00000			0.00000
 2			0.00000			0.00000			1.00000
 1			0.00000			1.00000			0.00000
 2			0.00000			0.00000			1.00000
 2			0.00000			0.00000			1.00000
 1			0.00000			1.00000			0.00000
 2			0.00000			0.00000			1.00000

All the predictions are correct in the result above

In [75]:
# iii. Report on the score for kNN, what does the score measure?
train_score = KNN_model.score(X_train, y_train)
test_score  = KNN_model.score(X_test, y_test)
print("Training score for model :", train_score)
print("Test score for model :", test_score)

Training score for model : 0.9629629629629629
Test score for model : 1.0


### Score:
The score of the model is between 0 to 1 (0% to 100%) and it represent the accuracy on the given data. The score of training data is 0.96, means the accuracy on the model is about 96%. The score of test data is 1.0, means model is perfectly accurate.

## Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Logistic Regression:
- Configuration: default
- Training Score:0.9703703703703703
- Test Score:1.0

Support Vector Machine:
- Configuration: default
- Training Score: 0.9703703703703703
- Test Score:1.0

Neural Network:
- Configuration: max_iter = 200
- Training Score: 0.9777777777777777
- Test Score:1.0

- Configuration: default
- Training Score: 0.9703703703703703
- Test Score:1.0

K-Nearest Neighbor:
- Configuration: default
- Training Score: 0.9629629629629629
- Test Score:1.0


Based on the result above, all model has test score 1.0. 

Neural Network performed the best on the dataset, because it has highest training score compare with other models.

I'm surprise that the test score of all model are 1.0, which means they all perfect and accurate. And im also suprise that the K-Nearest Neighbor model has 100% certain prediction on species.
